[Profile] Implement Payment Information section

ncksllvn opened the pull request:
Tested using User 32

```
npm run watch -- --entry=profile-360,auth,static-pages
```

Todo
- [x] Rename `paymentInformationStatus` to something like `paymentInformationModalState`
- [x] Move form status out of state and into store
- [x] Some level of unit/e2e tests

---

## Description
This PR implements a Payment Info section into the Profile app. Payment info is retrieved and saved via an EVSS service called PPIU. This work is is feature flagged to not render in Production.

Ticket - https://github.com/department-of-veterans-affairs/vets.gov-team/issues/18163

## Testing done
- Local testing

## Screenshots

### Profile
![image](https://user-images.githubusercontent.com/1915775/57255322-77f09d00-7021-11e9-9542-46a719b7c99f.png)

### Edit clicked
![image](https://user-images.githubusercontent.com/1915775/57255450-c0a85600-7021-11e9-8841-f8a0406e78eb.png)



## Acceptance criteria
- [x] Foundation for direct deposit functionality built to Staging

## Definition of done
- [x] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

Commit:
Initial commit

Commit:
Add reducer

Commit:
Add the view layer

Commit:
tweak

Commit:
Add edit modal

Commit:
Add the form

Commit:
Move gate to outer scope

Commit:
add handling for saves

Commit:
Add error-handling

Commit:
Undo changes to VAProfileApp

Commit:
Move isEditing into props

Commit:
Change status to uiState

Commit:
Move edit-modal state into store

Commit:
Merge remote-tracking branch 'origin/master' into payment-info-18163

Commit:
Add unit test for PaymentInformation container

Commit:
add unit test for edit modal

Commit:
Move modal props from reducer to modal

Commit:
add actions test

Commit:
add reducer test

Commit:
set min account num length to 6 chars

Commit:
update unit tests to require 6 chars

Commit:
update unit test acct number

Commit:
switch back to 1-digit acct numbers

ncksllvn commented on src/applications/personalization/profile360/tests/actions/paymentInformation.unit.spec.js:
```diff
@@ -0,0 +1,61 @@
+
+const teardown = () => {
+  global.fetch = oldFetch;
+};
```
Out of scope for this ticket, but it's such a widely-used pattern to mock `fetch` this way that we should probably centralize or create a helper for it.

ncksllvn commented on src/applications/personalization/profile360/components/ProfileView.jsx:
```diff
@@ -84,6 +85,7 @@ class ProfileView extends React.Component {
                 militaryInformation={militaryInformation}
               />
               <ContactInformation />
+              <PaymentInformation />
```
`PaymentInformation` requires the service `evss_claims` for it to be relevant to the user, which is a unique requirement from the other components rendered here. So, I decided to make `PaymentInformation` a container instead, so that I didn't have to pass along more props down the line.

erikphansen replied:
cool. this is the first time I've really looked at the `ProfileView` code, but I've noticed that on the Dashboard we are kind of all over the place with regard to determining which components we render; sometimes the logic is abstracted away in the component itself, sometimes the Dashboard itself knows if it should show or hide the subcomponent. It would be nice to be consistent, but I'd need to think a lot more about the ramifications of that.

ncksllvn replied:
Yeah, the Personalization apps went through a lot of transformation in a short time period. The Profile completely completely changed as we switched from PCIU to the transaction-based Vet360. They have been cleaned up significantly since then but obviously they aren't perfect.

ncksllvn replied:
In some follow-up work for this, I'll try to clean the Profile a bit, but I don't have the bandwidth right now.

erikphansen replied:
I wouldn't worry about that. It's not a priority, IMO, more of a "fix stuff as you go along" kind of thing.

erikphansen reviewed (commented):
still trying to fully understand how the modal component works but wanted to share a handful of initial thoughts!

ncksllvn commented on src/applications/personalization/profile360/containers/PaymentInformation.jsx:
```diff
@@ -0,0 +1,148 @@
+
+    if (this.props.isLoading) {
+      return <LoadingIndicator message="Loading payment information..." />;
+    }
```
There will be a couple other layers here in a follow-up PR - the "add" state for when the user is eligible for direct deposit but has not yet set it up, and we'll also be enforcing MFA to use this tool.

erikphansen replied:
MFA?

ncksllvn replied:
Ah yeah I mean 2FA (text-message account verification during login configured in ID.me)

ncksllvn commented on src/applications/personalization/profile360/actions/paymentInformation.js:
```diff
@@ -0,0 +1,58 @@
+    type: SET_PAYMENT_INFO_UI_STATE,
+    state,
+  };
+}
```
I wrote this action as sort of a replacement for `setState`, since the vets-website pattern is to keep most state in the store as opposed to component state.  It's used to pass data like the form fields, `isEditing, `isSaving`, etc.

jbalboni replied:
Our guidance is, generally, that action names should reflect something a user did or an external event. Meaning, a user didn't try to set the payment information ui state, they opened the edit view of the modal or saved something.

ncksllvn replied:
Updated to use dedicated actions. This makes sense to me now that we're encouraging deriving state in the reducer, although I still think we can improve by setting guidelines for `setState`.

ncksllvn commented on src/applications/personalization/profile360/components/PaymentInformationEditModal.jsx:
```diff
@@ -0,0 +1,228 @@
+    );
+  }
+}
+
```
The forms system didn't feel like a good fit for a minimal form like this, but my experience with that library is pretty slim. What do you think?

erikphansen replied:
Ah I was gonna ask about that. on the new ID form page that logged-out users see before they can fill out the HCA, I used a `SchemaForm` component directly. I'm not sure if something like that would be useful here as well. https://github.com/department-of-veterans-affairs/vets-website/blob/01f348a9b64ce9039c91534e5bd7c754fb64f337/src/applications/hca/containers/IDPage.jsx#L180

erikphansen replied:
It might be nice to use SchemaForm here as well. But half the reason to do so is to avoid the work of setting up each form field directly and you already did that :-p

jbalboni replied:
Yeah, this form is tiny, I think this approach or using SchemaForm directly seems fine. The advantage with SchemaForm is that you don't have to write a bunch of focus and validation handling yourself, you can use our well-tested forms code.

For example, the `onSubmit` method in here needs to set the focus to the first field with an error when there are outstanding errors.

ncksllvn replied:
I updated `onSubmit` to set focus. I will consider switching to `SchemaForm` in a follow-up ticket but I don't think it's necessary at this time.

ncksllvn commented on src/applications/personalization/profile360/components/PaymentInformationEditModal.jsx:
```diff
@@ -0,0 +1,228 @@
+            We’re sorry. Something went wrong on our end and we couldn’t save
+            the recent updates you made to your profile. Please try again later.
+          </p>
+        </AlertBox>
```
We'll be adding error-handling for special cases in a follow-up ticket.

ncksllvn commented on src/applications/personalization/profile360/components/PaymentInformationEditModal.jsx:
```diff
@@ -0,0 +1,228 @@
+  onClose = () => {
+    this.props.onClose();
+    this.setEditedState(PaymentInformationEditModal.defaultEditModalFields);
+  };
```
Here ,the `defaultEditModalFields` are added into the store under the `vaProfile.paymentInformationUiState.editModalFields` to initialize this component's props.

ncksllvn reviewed (commented):
Added a few comments to help a bit

ncksllvn commented on src/applications/personalization/profile360/reducers/index.js:
```diff
@@ -23,6 +36,50 @@ function vaProfile(state = initialState, action) {
+      return {
+        ...state,
+        paymentInformation: action.paymentInformation,
+        paymentInformationUiState: {
```
It also contains `editModalFields` - these [values](https://github.com/department-of-veterans-affairs/vets-website/pull/10028/files#diff-ef09c9ed9cd0cf7fb86ad08b406c4097R22) are added into the store during the `componentDidMount` of the edit-modal. Typically I think React apps would define this in the constructor under component state, but our pattern tends to be adding as much as we can into the store. What do you think?

ncksllvn commented on src/applications/personalization/profile360/actions/paymentInformation.js:
```diff
@@ -0,0 +1,58 @@
+import { apiRequest } from 'platform/utilities/api';
+import { getData } from './index';
+
+export const FETCH_PAYMENT_INFORMATION_SUCCESS =
```
Ah interesting, I remember that consensus but I must have misread it back then. I'll update that.

jbalboni replied:
This is written up here: https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/platform/front-end-standards/redux

ncksllvn replied:
Updated to use past-tense

ncksllvn commented on src/applications/personalization/profile360/actions/paymentInformation.js:
```diff
@@ -0,0 +1,58 @@
+
+      dispatch({
+        type: SAVE_PAYMENT_INFORMATION_SUCCESS,
+        paymentInformation: response.data.attributes,
```
Sure!

ncksllvn commented on src/applications/personalization/profile360/containers/PaymentInformation.jsx:
```diff
@@ -0,0 +1,148 @@
+      return <LoadingIndicator message="Loading payment information..." />;
+    }
+
+    const { paymentAccount } = this.props.paymentInformation.responses[0];
```
This is good call, and this will definitely be addressed once we work on the "add" state for direct deposit - users who are eligible but have not set it up yet - because right now I don't know exactly what that empty state will look like. I think it's possible `paymentInformation` will be `null` or an empty array, but based on some API discovery it seems most likely that we'll just see a standard `paymentAccount` data structure with empty values.

ncksllvn replied:
This will be addressed once we have more insight into what errors/null fields look like from PPIU.

Commit:
export account_type_options

jbalboni commented on src/applications/personalization/profile360/actions/index.js:
```diff
@@ -8,7 +8,7 @@ export const FETCH_MILITARY_INFORMATION_SUCCESS =
   'FETCH_ADDRESS_CONSTANTS_SUCCESS';
 
-async function getData(apiRoute) {
+export async function getData(apiRoute) {
```
It doesn't feel like this should be exported from an actions file. It's a helper function, if we need it in multiple places, maybe it makes sense in the api utilities?

ncksllvn replied:
Moved to utils

jbalboni commented on src/applications/personalization/profile360/components/PaymentInformationEditModal.jsx:
```diff
@@ -0,0 +1,228 @@
+    paymentInformationUiState: PropTypes.object,
+  };
+
+  static defaultEditModalFields = {
```
Why not put this in the initial state of the Redux store?

ncksllvn replied:
I need to reset the form after closing/opening the modal so I kept these values close by to do that, but if we're in favor of moving validation into the reducer, then it makes sense to me to do that reset in the reducer as well.

ncksllvn replied:
Added to the initial state

Commit:
Change action names to be consistent

jbalboni commented on src/applications/personalization/profile360/components/PaymentInformationEditModal.jsx:
```diff
@@ -0,0 +1,228 @@
+    }
+  };
+
+  onRoutingNumberChanged = field => {
```
I think this method is an example of where we sometimes don't take advantage of Redux. This is all data/state logic that you'd expect to see in a reducer in an app that's using Redux. In there, this logic would be plain JS that's pretty easy to test (and easier to share between multiple fields). And this component would get simpler and be focused on displaying the right information and dispatching the right action depending on user input.

erikphansen replied:
agreed that this should be put somewhere where it's easy to test in insolation. but still feeling like just a beginner with Redux, I'm not sure how putting this logic in the reducer would work. I'm guessing if I dug deep enough into the form system that question would be answered since that must be how the `ui:validations` work.

jbalboni replied:
It is, but there's a lot of extra logic around it in the forms system, so it'll be hard to figure out that way.

But essentially, you would have the React component fire an action with the info you need:

```js
onRoutingNumberChanged = field => {
  this.props.updateFormField('financialInstitutionRoutingNumber, field);
}
```
In your action creator file:

```js
function updateFormField(fieldName, field) {
  return {
    type: UPDATE_FORM_FIELD,
    fieldName,
    field
  }
}
```

And then in the reducer:

```js
case UPDATE_FORM_FIELD:
  // set from our data utilities
  return set(`paymentInformationUiState.${action.fieldName}`, {
    value: {
      ...action.field,
      // could also theoretically leave this in component, but you could write a function like this as well
      errorMessage: getErrorMessage(action.fieldName, action.field.value),
    }
  };
```

ncksllvn replied:
Yeah, this is my first loop back to the React world and maybe this is standardized now, but I expected criticism if I derived these values in the reducer, because they could be considered side effects. I would prefer to have these in the reducer though.

erikphansen replied:
Thanks for that example, @jbalboni. Seeing that the `value` of each form field is an object that wraps up the "real" value along with an error message is both a "duh" and "aha" realization for me. Would the `getErrorMessage()` helper also be responsible for only validating if the field was dirty?

ncksllvn replied:
That works for me @jbalboni!

jbalboni replied:
I don't think anything in here would be considered side effects. Having side effects would mean changing something other than the Redux state in a reducer, which isn't the case. The other way you could violate a "rule" in Redux reducers is to have them use some code that isn't solely dependent on the arguments passed into the reducer (i.e. use moment to get today's date). That's not happening here either, from what I saw.

Sometimes people are also strict about making the state held in Redux very minimal, but there's no hard and fast rule there.

jbalboni replied:
@erikphansen It could, but actually, I prefer always running the validation and then having the React components determine if a validation message is shown based on that dirty field. The forms system does it that way (though it doesn't keep validation state in Redux at all).

ncksllvn replied:
I moved the form values into the reducer, while validation happens in the component at during  `onValueChange` . LMK if it still needs improvement though.

Commit:
Move to reducer

Commit:
apply focus

Commit:
Move getData to utils

Commit:
Rm bad commit

Commit:
Cleaning up

Commit:
Move gate to featureFlags.js

Commit:
Run validation again on submit

ncksllvn commented on src/applications/personalization/profile360/components/PaymentInformationEditModal.jsx:
```diff
@@ -0,0 +1,228 @@
+import ErrorableTextInput from '@department-of-veterans-affairs/formation-react/ErrorableTextInput';
+import ErrorableSelect from '@department-of-veterans-affairs/formation-react/ErrorableSelect';
+
+const ACCOUNT_TYPES_OPTIONS = {
```
exported

ncksllvn commented:
I need to update and write additional the unit tests which I'm doing now, but if possible I would like to have this merged today because I'm out Mon/Tues and this work needs to move along. @jbalboni @erikphansen if you have time PTAL. I think I addressed most of your points.

Commit:
update actions tests

Commit:
update reducer tests

Commit:
Update container tests

Commit:
Update edit-modal tests

Commit:
Add tests for utils

Commit:
Account for empty state just in case
